src/basic_recipes/raincloud.jl

```diff
@@ -0,0 +1,282 @@
+    base_min, base_max = (-0.5, 0.5)
+    jitter = [rand_localized(base_min, base_max) for _ in 1:length_data_array]
+
+    # created clamp_min, and clamp_max to clamp a portion of the data
+    @assert (base_max - base_min) == 1.0
+    @assert (base_max + base_min) / 2.0 == 0
+    clamp_min = base_min + (clamped_portion / 2.0)
+    clamp_max = base_max - (clamped_portion / 2.0)
+
+    # clamp if need be
+    clamp!(jitter, clamp_min, clamp_max)
+
+    # Based on assumptions of clamp_min and clamp_max above
+    jitter = jitter * (0.5jitter_width / clamp_max)
+
+    return jitter
+end
+
+####
+#### Functions that make the cloud plot
+####
+function plot!(
+        ax::Makie.Axis, P::Type{<: RainClouds},
+        allattrs::Attributes, category_labels, data_array)
+
+    plot = plot!(ax.scene, P, allattrs, category_labels, data_array)
+
+    if eltype(category_labels) <: AbstractString
+        ulabels = unique(category_labels)
+        ax.xticks = (1:length(ulabels), ulabels)
+    end
+    if haskey(allattrs, :title)
+        ax.title = allattrs.title[]
+    end
+    if haskey(allattrs, :xlabel)
+        ax.xlabel = allattrs.xlabel[]
+    end
+    if haskey(allattrs, :ylabel)
+        ax.ylabel = allattrs.ylabel[]
+    end
+    reset_limits!(ax)
+    return plot
+end
+
+function group_labels(category_labels, data_array)
+    grouped = Dict{eltype(category_labels), Vector{Int}}()
+    for (label, data_ix) in zip(category_labels, axes(data_array,1))
+        push!(get!(grouped, label, eltype(data_array)[]), data_ix)
+    end
+
+    return pairs(grouped)
+end
+
+function ungroup_labels(category_labels, data_array)
+    if eltype(data_array) isa AbstractVector
+        @warn "Using a nested array for raincloud is deprected. Read raincloud's documentation and update your usage accordingly."
+        data_array_ = reduce(vcat, data_array)
+        category_labels_ = similar(category_labels, length(data_array_))
+        ix = 0
+        for (i, da) in enumerate(data_array)
+            category_labels_[axes(da, 1) .+ ix] .= category_labels[i]
+            ix += size(da, 1)
+        end
+        return category_labels_, data_array_
+    end
+    return category_labels, data_array
+end
+
+function convert_arguments(::Type{<: RainClouds}, category_labels, data_array)
+    cloud_plot_check_args(category_labels, data_array)
+    return (category_labels, data_array)
+end
+
+
+function plot!(plot::RainClouds)
+    category_labels = plot.category_labels[]
+    data_array = plot.data_array[]
+    category_labels, data_array = ungroup_labels(category_labels, data_array)
+    if any(ismissing, data_array)
+        error("missing values in data not supported. Please filter out any missing values before plotting")
+    end
+
+    # Checking kwargs, and assigning defaults if they are not in kwargs
+    # General Settings
+    # Define where categories should lie
+    x_positions = if eltype(category_labels) isa String

```

Comment:
Whoops, sorry, looks like this needs to change. Sorry, I forgot to run the doctests locally.

```suggestion
    x_positions = if eltype(category_labels) isa AbstractString
```